### PR TITLE
ui: Ember upgrade 2.18 > 3.8

### PR DESCRIPTION
**Pull request by johncowen**
~WIP: Detailed description to come, please see commit comments in the meantime.~

~Please note, all tests apart from lint tests are passing, we still have to go through and remove all the ununsed `import { get } from '@ember/object'` that we don't use anymore.~

We were hoping to upgrade straight to 3.14 but the time came to do this so we've gone for 3.8 instead, another similar upgrade should follow as soon as 3.14 is out.

This PR is mostly a run of `ember-cli-update` plus the associated code-mods.

Various things also broke as a result of the upgrade, they are pretty well split out into individual commits but I'll detail the interesting ones here:

1. Previously we were dynamically adding methods to new instances of Proxies (used for EventSources to be invisibly passed through Routes) on demand, you can no longer do this and call `this._super`. Therefore we alter how the proxies are created by creating classes of Proxies on demand and caching and then creating instances from those. (1d55c986513135a64d90896bef820a517cebefae)

2. We had an interesting issue were although the UI worked fine, the ember tests would complain about `this` being referenced before `super`. We noticed that `this` wasn't being referenced before `super` in the code, but once transpiled it probably would be due to the transpiled arrow function placing the `this` reference before `super` to then be referenced from another scope. We also hit an issue were you can't `bind/apply` to the constructor of a class. To solve all this we reverted the classes involved back to ES5 class syntax which solved everything. Lastly, there's potentially a tiny refactor here that I've already kind of started to not depend on access to `this`, the PR here is only to upgrade though and as we could get away without changing how things worked we did (a240ecd745a837c39550f343335b1d9398f9142c)

3. Similar to point 1 here, it looks like you can no longer just hang `computed` properties off of new instances like you used to be able to. We were using this ability to add an `isSaving` state to our `Changeset` usage (`ember-changeset`). We used a similar approach to point 1 and added it to the class instead - future work will probably move this to an initializer (e33db406cefa8774f483d87b1f2e13fc0349cb89)

4. New ember API's have changed how you can detect whether something is a computedProperty or not. It's not immediately clear if its even possible now. Therefore we change how we detect whether something should be listened to or not by just looking for presence of `addEventListener`. I'm not massively happy with this as I'd rather continue to look to see if something is already defined as a computedProperty or not, so if anyone has any hinters here it would be much appreciated. For now at least, this does the job (9e33e52e335fb95f6b08bcc8e064b707a36b8914)

5. We skipped one test. This test is testing API that we don't actually use, and seems to be as a result of a change to the timing of `RSVP` in the upgrade. It doesn't affect the UI itself as far as we can see (it would if it was quicker than before the upgrade, but in this case its slower than before the upgrade). I had a quick look into it and it actually seems like the previous behaviour doesn't make sense rather than the current behaviour, so while its something I'd like to look into more I've just skipped the test for the moment. The 'fix' here might be to change how the test works rather than fix it, as bad as that sounds (due to the comment that the previous behaviour doesn't seem to make sense). Again not massively happy with this one, but happy to move on without it for the moment (9e33e52e335fb95f6b08bcc8e064b707a36b8914).

6. We deleted one test. This test was an 'alarm' test for a private function we were using to tell us if the code for that function ever changed. We upgraded the package it is from, hence the alarm tripped. We no longer use this code anyway (since https://github.com/hashicorp/consul/pull/5637), so the alarm test is also no longer necessary (a17f082f09ad57571537057dcfe83cbf808ee563)

^ those aren't all the changes we had to make for the upgrade, but apart from https://github.com/hashicorp/consul/pull/5637 they are the most interesting ones, please see commits themselves for more info.

### Todo

- [x] Remove all `get` imports that are no longer required due to the proxy upgrade (and the code mod changing most of these for us)
- [ ] ~Go through the code to see if the code mods have missed any places where they should have changed from `get` to just normal property access (thanks @backspace !) and also check they haven't been removed from places where we used `get` for safe deep property access(`get('a.property.than.might.exist.but.doesnt.but.dont.throw.an.error'`)~ - to be done in a later PR
- [x] Final checks/cleanup/address any left over code TODO's
- [x] Possibly, uninstall `ember-cli-upgrade` from the `package.json` file
- [ ] ~Possibly run a last `yarn upgrade` on other dependencies (potentially do this in another PR)~ to be done in another PR (reasonably soon)

### CODE REVIEW

File: ui-v2/app/mixins/with-event-source.js [hunk 1]
```diff
@@ -33,7 +44,10 @@ export default Mixin.create(WithListeners, {
           // setProperties will be called the next time we enter the Route so this
           // is ok for what we need and means that the above conditional works
           // as expected (see 'here' comment above)
-          delete this[prop];
+          // delete this[prop];
+          // TODO: Check that nulling this out instead of deleting is fine
```
Comment (backspace) [inline ui-v2/app/mixins/with-event-source.js]: Is this TODO for this PR or for the future?

File: ui-v2/app/routes/dc/acls/tokens/index.js [hunk 2]
```diff
@@ -13,26 +13,24 @@ export default Route.extend(WithTokenActions, {
...
-        if (token && get(token, 'AccessorID') === null) {
-          // If you return here, you get a TransitionAborted error in the tests only
-          // everything works fine either way checking things manually
-          this.replaceWith('dc.acls');
-        }
-      });
+    return this.settings.findBySlug('token').then(token => {
+      // If you have a token set with AccessorID set to null (legacy mode)
+      // then rewrite to the old acls
+      if (token && get(token, 'AccessorID') === null) {
```
Comment (backspace) [inline ui-v2/app/routes/dc/acls/tokens/index.js]: This `get` could be converted too, right?

File: ui-v2/lib/block-slots/addon/components/block-slot.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { readOnly } from '@ember/object/computed';
 import Component from '@ember/component';
 import { isPresent } from '@ember/utils';
 import { get, set, computed, defineProperty } from '@ember/object';
```
Comment (backspace) [inline ui-v2/lib/block-slots/addon/components/block-slot.js]: It seems like `computed` is no longer used? Also, `get` is still present on line 42, but it could probably be removed.

File: ui-v2/tests/unit/controllers/dc-test.js [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-// Replace this with your real tests.
-test('it exists', function(assert) {
-  let controller = this.subject();
-  assert.ok(controller);
+  // Replace this with your real tests.
+  test('it exists', function(assert) {
+    let controller = this.owner.lookup('controller:dc');
+    assert.ok(controller);
+  });
 });
```
Comment (backspace) [inline ui-v2/tests/unit/controllers/dc-test.js]: I’d probably do it in another PR but it seems like you might as well remove this and the other placeholder controller tests! They’re pretty passé 😆

File: ui-v2/tests/unit/mixins/click-outside-test.js [hunk 5]
```diff
@@ -2,11 +2,11 @@ import EmberObject from '@ember/object';
...
-  let subject = ClickOutsideObject.create();
-  assert.ok(subject);
+module('Unit | Mixin | click outside', function() {
+  // Replace this with your real tests.
+  test('it works', function(assert) {
+    let ClickOutsideObject = EmberObject.extend(ClickOutsideMixin);
+    let subject = ClickOutsideObject.create();
+    assert.ok(subject);
+  });
 });
```
Comment (backspace) [inline ui-v2/tests/unit/mixins/click-outside-test.js]: Might as well remove these blueprint-y mixin tests too?

File: ui-v2/tests/unit/routes/dc-test.js [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
+  setupTest(hooks);
 
-test('it exists', function(assert) {
-  let route = this.subject();
-  assert.ok(route);
+  test('it exists', function(assert) {
+    let route = this.owner.lookup('route:dc');
+    assert.ok(route);
+  });
 });
```
Comment (backspace) [inline ui-v2/tests/unit/routes/dc-test.js]: Also these route tests… I pretty much never see them in recent Ember applications.

File: ui-v2/tests/unit/services/atob-test.js [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
 
-// Replace this with your real tests.
-test('it exists', function(assert) {
-  let service = this.subject();
-  assert.ok(service);
+  // Replace this with your real tests.
+  test('it exists', function(assert) {
+    let service = this.owner.lookup('service:atob');
+    assert.ok(service);
+  });
```
Comment (backspace) [inline ui-v2/tests/unit/services/atob-test.js]: same for service tests… I do find these useful on occasion, but there doesn’t seem to me to be much point in having the “it exists” ones.

File: ui-v2/tests/unit/utils/http/request-test.js [hunk 8]
```diff
@@ -1,10 +1,10 @@
...
-// Replace this with your real tests.
-test('it works', function(assert) {
-  const actual = httpRequest;
-  assert.ok(typeof actual === 'function');
+module('Unit | Utility | http/request', function() {
+  // Replace this with your real tests.
+  test('it works', function(assert) {
+    const actual = httpRequest;
+    assert.ok(typeof actual === 'function');
+  });
```
Comment (backspace) [inline ui-v2/tests/unit/utils/http/request-test.js]: Another placeholder test 🎉

File: ui-v2/.travis.yml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    # See https://git.io/vdao3 for details.
+    - JOBS=1
+
+before_install:
+  - npm config set spin false
+
+script:
+  - npm run lint:hbs
+  - npm run lint:js
+  - npm test
```
[the remark this answers is not in the document]
Comment (johncowen) [reply ui-v2/.travis.yml]: hehe ah yes, we've never run any `ui-v2` tests in travis! Gonna delete this 👍

File: ui-v2/README.md [hunk 10]
```diff
@@ -61,6 +61,10 @@ Please note: You do not need to run `make start-api`/`yarn run start:api` to run
 * `make test` or `yarn run test`
 * `make test-view` or `yarn run test:view` to view the tests running in Chrome
 
+### Linting
+
+### Building
+
```
[the remark this answers is not in the document]
Comment (johncowen) [reply ui-v2/README.md]: I'll see if I can fill them out with something useful, if not I'll remove them, thanks!

File: ui-v2/app/mixins/with-event-source.js (hunk 1 above)
Comment (johncowen) [reply ui-v2/app/mixins/with-event-source.js]: Ah yeah I meant to ask about this. I'm sure I remember seeing an ember `unset` or similar, but I can't find anything (so something that deletes a property entirely not just sets it to null). Does that ring any bells with you?

I think I tried doing `set(this, prop, undefined)` here also, maybe thats the 'sanctioned' way of completely unsetting a property? In the end I settled with `null`, as far as I remember when I looked where we were relying on the `undefined` check, a `null` check instead seemed just as good.

If you know what the sanctioned way to completely unset a property is in a 3.8 world, that would be really good to know.

I'm gonna relook at the code surrounding this in a sec, just to confirm what I've done here is 'fine' (or have another look at doing a real unset, which I'd prefer), then the TODO will go in this PR.

File: ui-v2/app/routes/dc/acls/tokens/index.js (hunk 2 above)
Comment (johncowen) [reply ui-v2/app/routes/dc/acls/tokens/index.js]: Hmm yeah not sure, I wonder why the codemod skipped this?

This is a `null` check, not an undefined/falsey check, maybe thats why?

Thanks for picking up on this, I'll have a little look and see if I can understand why the codemod skipped it. I'm a bit cautious about just changing it without understanding why it was skipped, I would assume theres a reason behind it? I don't suppose you can think of any other reasons it might skip this?

Anyway will come back here once I've had a little look into it

**Comment by johncowen** [comment]
@alvin-huang just a little heads up here incase.

I hit a little issue with the ember-build CI script not working ( see https://github.com/hashicorp/consul/pull/6317/commits/5a826c283b0ad67aa293e5b001b4948a4c6888a3 and https://github.com/hashicorp/consul/runs/199119980 ), and I noticed it wasn't going through the same path as the other tooling scripts (Makefile > install deps if they aren't already there > yarn script) so I changed them to use that.

Looking at it now it seems like it was some sort of flake, not sure exactly, but in any case as the Makefile checks to see if the deps and only installs them if they aren't (via yarn), it won't effect anything leaving this commit in, and would prevent the ember-build from failing for whatever reason it can't find the cached dependencies from ember-cache.

Feel free to ping me offline if ^ isn't clear enough and you want more info - I think all in all it's 'nothing to see here, please move along' but thought I'd ping you just in case.

File: ui-v2/lib/block-slots/addon/components/block-slot.js (hunk 3 above)
Comment (johncowen) [reply ui-v2/lib/block-slots/addon/components/block-slot.js]: Hmmm yeah this one is really weird..

2 bits of weirdness:

A codemod added the `{ readOnly }`, but it seems like the `get` codemod missed/skipped the `get` here. I'm just double checking to make sure there's no reason it missed/skipped it before changing it manually.

---

I've also noticed that ember/testem eslint tests don't seem to include any files in `lib`. (my custom linter does pick up on this though, which uses the same .eslintrc file)

Do you know if this is default ember configuration? And if so any idea where I can make sure ember includes any files in `lib` during testem testing?

**Comment by johncowen** [comment]
Hey @backspace

Thanks again for giving this the early once over, much appreciated. There are a few things above I'm still looking into and also I left a few queries above you may have more knowledge on than me, but I just wanted to comment on the placeholder tests in one place here instead of inline.

Basically when I started on this 18 months or so ago, I'd not used ember before and I put a lot of effort into trying to stay on the ember 'cowpath'. In my head that involved using ember's generators for everything. As you probably know in ember 2.18 the generators also make you those placeholder tests, which only really test whether a class/component or whatever can be instantiated.

Back then I was very much like 'do you really need to keep these in here?' And my answer to myself at the time was:

1. This is what ember does for you, so stick to it. (I'm actually interested to see what generators in 3.8 do for you)
2. It was more effort to remember to delete them than to just commit them in.
3. They do no harm.
4. When I was starting with ember it was really difficult to remember where certain packages lived `@ember/computed`, `@ember/object/computed`, `ember-computed`, `ember/computed` etc etc etc. Most of ember itself lives in `node_modules/ember-source` (and some other bits elsewhere), and there are no clues as to where things live by just looking in `node_modules` . I used to add features with my test suite permanently open in a window to the side, and therefore one of these placeholder 'instantiation' tests would immediately fail in the corner of my eye if I'd tried to import something and got it completely wrong. So, back then, for me these were useful. Although not so much now I know where everything lives :).

All in all, personally I'd be happy to get rid of them now also, my only reason to maybe keep them is if someone else comes along with as little ember knowledge as I had (potentially possible real soon) and also finds these as vaguely useful as I did 18 months ago.

Whichever we go with, I completely agree, it should be done in another PR.

Lastly, still looking into a few of your above comments, will ping you again once I'm done with all that - I might ping you offline to chat in a bit.

File: ui-v2/app/mixins/with-event-source.js (hunk 1 above)
Comment (backspace) [reply ui-v2/app/mixins/with-event-source.js]: I think `delete` is a fine way of accomplishing this. I don’t think this has changed in Ember in years, at least. You can `set` to `undefined` too, it might have some slight differences in behaviour depending on the context… I guess the thing to do is see how it actually acts 😯

File: ui-v2/app/routes/dc/acls/tokens/index.js (hunk 2 above)
Comment (backspace) [reply ui-v2/app/routes/dc/acls/tokens/index.js]: I once wrote a codemod and when things appear in different contexts, they sometimes have to be specifically enumerated. So maybe because this is in an expression…? I’m not sure, but I can’t see any reason this can’t be converted, I think it’s more an instance of how difficult it is to cover all the cases.

File: ui-v2/lib/block-slots/addon/components/block-slot.js (hunk 3 above)
Comment (backspace) [reply ui-v2/lib/block-slots/addon/components/block-slot.js]: hmm well test-loading _used_ to happen with [this addon](https://github.com/ember-cli/ember-cli-test-loader) I believe, but it seems like it might not be used anymore. I don’t know how to get it to run tests outside `tests/`, I guess I always follow the conventions in this way.

**Comment by johncowen** [comment]
Additional note here while I remember to add it, we now get a few deprecation warnings in `console` whilst running the app. These will be looked at in a later PR (if a `yarn upgrade` of the left over dependencies doesn't sort them).

For example:

Screenshot 2019-08-21 at 15 58 29

File: ui-v2/app/mixins/with-event-source.js (hunk 1 above)
Comment (johncowen) [reply ui-v2/app/mixins/with-event-source.js]: Yeah its weird, I couldn't remember the exact error doing this fixes, so I checked back in the commits. Took me a while to find it again but its this:

Screenshot 2019-08-21 at 16 44 24

^ that's what happens when you just use vanilla `delete` here. As soon as I change it to use `set`, setting it to `null` or `undefined` I don't get the error anymore. The error message is a little weird and sounds unrelated, but changing the `delete` to something `set` based solves it - and as this is all proxy related it probably makes sense that this fix and the error message are related.

I'm quibbling about using `undefined`, it feels a bit wrong setting things to `undefined`, which is why I chose `null` here instead. When I re-read the code it still achieves the same thing with `null` so I left it at that.

I 'think' I'm gonna leave this as is and remove the TODO, I've had a good look into it and I think this is the best I'm gonna get here. Unless you can think of anything else we can try?

Comment (backspace) [reply ui-v2/app/mixins/with-event-source.js]: Setting to `undefined` is more equivalent to using `delete`, but if it works the same way with either, it doesn’t really matter to me!

Comment (johncowen) [reply ui-v2/app/mixins/with-event-source.js]: Cool thanks @backspace I went with `undefined` in the end. The aim here is to change as little as possible aprt from the upgrade, and you're right, setting to `undefined` is closer to a delete.

I also added some more detailed notes here just incase.

Of further note - this is part of the code that will probably go once we add the work we've been doing on our `<DataSource />` component, as none of it will be required anymore - so as long as it gets us over the line with the upgrade, and we can start using `<DataSource />` its not a biggie in the scheme of things.

Thanks for the help and convo!

File: ui-v2/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (johncowen) [reply ui-v2/README.md]: I've added something more useful to these, weirdly you can only see the addition in certain areas of the GH UI, but they are there! ✅

File: ui-v2/app/routes/dc/acls/tokens/index.js (hunk 2 above)
Comment (johncowen) [reply ui-v2/app/routes/dc/acls/tokens/index.js]: Hey @backspace 

I've just taken a bit of a closer look through the source code and there are (at a rough estimate) over 100 places where the codemod has left in the `get`. I think some of them are valid (as usage for a safe deep property accessor), but others probably could be switched out.

Right now I've decided I'm a little reluctant to change them in this PR, reasons being:

1. `get` is still perfectly valid to use (although admittedly I'd rather not)
2. There is still a fair amount of work/approval to get:
  - https://github.com/hashicorp/consul/pull/6317
  - https://github.com/hashicorp/consul/pull/5637
  - https://github.com/hashicorp/consul/pull/6285
.. finally merged into `master`. I'd rather leave this for another day and concentrate on getting the rest over the line.
3. I'm trying to avoid changing any behaviour without realising - ie. I want to get the upgrade done with as little alterations as possible (there are already a fair few! 😄 ), right now tests pass, and the application looks good.
4. I'd like to understand why certain `get`s have been left before replacing them.

TL;DR I'd like to convert all these `get`s just not in this PR 😄 

Lemme know what you think.

File: ui-v2/lib/block-slots/addon/components/block-slot.js (hunk 3 above)
Comment (johncowen) [reply ui-v2/lib/block-slots/addon/components/block-slot.js]: I took another brief look at this, but couldn't figure out how I can get these files included into my ember/testem test suite, so gave up. I do have a `make lint` Makefile target that runs a linter on all of these additional files (actually I could just add this before my `make test`, not in this PR though). This is a manual thing, but as its a linter, I'm ok with that.

Once I'd given up on looking into this further, I also manually updated the things the codemod didn't get in this file.

File: ui-v2/app/components/healthchecked-resource.js [hunk 11]
```diff
@@ -1,15 +1,16 @@
+import { filter } from '@ember/object/computed';
```
Comment (johncowen) [inline ui-v2/app/components/healthchecked-resource.js]: Hmm, not massively sure I wanted it to do this 🤔 . I suppose its looking forwards to tree shaking, but I'm already importing computed for another thing further down, so I kinda like using the namespaced `computed.filter` 🤷‍♂ 

Not gonna change anything here, but was interesting nonetheless

File: ui-v2/app/routes/dc/acls/tokens/index.js (hunk 2 above)
Comment (backspace) [reply ui-v2/app/routes/dc/acls/tokens/index.js]: yes, I’m definitely in favour of minimal-intervention Ember version updates, so makes sense to me!

**Comment by johncowen** [comment]
Last comment from me here hopefully just to detail what is to be done after this PR, but is related to this PR:

1. Investigate the many usages of `get` that have been left over and see if we can convert them to use proxies.
2. There are a few places where the codemod has replaced instances of `get` with `this.prop` that looked like:

```javascript
const something = get(this, 'something');
this.useSomethingOnce(something).useItAgain(something);

// that was changed to

const something = this.something;
this.useSomethingOnce(something).useItAgain(something);
```

.. we may aswell just change all instances of this to

```javascript
this.useSomethingOnce(this.something).useItAgain(this.something);
```

It will be a large-ish change and there is no rush to do this, and doesn't effect the end user in the slightest, but is nice for a cleaner codebase.

3. `yarn upgrade` everything else now we are on a later version of ember 🎉

**Comment by johncowen** [comment]
Closing in favour of https://github.com/hashicorp/consul/pull/6448
